### PR TITLE
Add context-extension capability and merged-eg support to nnet3-chain-copy-egs 

### PR DESCRIPTION
**Pull request by hhadian**
The only thing remaining is to change the "truncate" code to support meged-egs. 
I thought you might have comments regarding how I'm doing the context extension, so I created the work-in-progress PR.

### CODE REVIEW

File: src/chainbin/nnet3-chain-copy-egs.cc [hunk 1]
```diff
@@ -224,41 +267,33 @@ void CalculateFrameSubsamplingFactor(const NnetChainExample &eg,
 void ModifyChainExampleContext(int32 left_context,
                                int32 right_context,
                                const int32 frame_subsampling_factor,
-                               NnetChainExample *eg) {
-  static bool warned_left = false, warned_right = false;
+                               NnetChainExample *eg,
```
Comment (danpovey) [inline src/chainbin/nnet3-chain-copy-egs.cc:280]: It's not clear to me what this function is supposed to do (or, actually, what it originally did).  Please add documentation!

**Comment by danpovey** [comment]
It seems to me that the mechanism for extending the left-context is quite different than the mechanism for removing left-context, and maybe it would be more elegant if the same code did all the context modification.  But I haven't read the code recently so maybe there are reasons why this might be hard to do.  Also I might in principle be OK with moving some code to the library dir.  But I don't have super strong opinions about any of this.

**Comment by hhadian** [comment]
Yes, they are quite different. I think it's more clear if the codes for addition/removal of context are separate.

**Comment by danpovey** [comment]
OK.  If you could try to run it on some egs on disk and inspect the result to make sure it's doing the right thing, that would be great.

**Comment by hhadian** [comment]
Will do.

**Comment by hhadian** [comment]
I looked at the result and they seem good.
I also tested this by comparing the output of `copy-egs | merge-egs` with that of `merge-egs | copy-egs` while changing the contexts and the outputs were identical.

**Comment by danpovey** [comment]
Thanks.  Merging.

**Pull request merged**